### PR TITLE
Manage internal and external dependencies

### LINKED ISSUE #1: Handling of internal and external dependencies of packages
[**Updated** 20190403]

We are of the opinion that we should use `^` where possible and sensible to ensure we run with the most recent updated dependencies. The lock file should be committed and updated at regular intervals.

We decided to use the `^` symbol or version ranges for "internal" dependencies. Internal dependencies are projects under the `@openst` project.

[ **Updated** 20190411]
For external dependencies we want to use fixed versions.

Until we are at version `1.0.0`, we should use a version range like `>=0.10.0 <0.11.0` to prevent propagation of breaking changes in newer minor versions.

The scope of this ticket is exclusively an update to the developer guidelines.

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR adds guidelines for managing internal and external dependencies.

Fixes #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -183,3 +183,25 @@ The PR is the only place any team member is expected to consult for information
 [soliumrc]: ./.soliumrc.json
 [travisci]: https://travis-ci.org/
 [webpack]: https://webpack.js.org/
+
+
+## Managing internal and external dependencies
+* Internal dependencies:
+    1. Internal dependencies are dependencies which are under `@openst` project. 
+    2. Until the stable version released, we should use version ranges to avoid any breaking change in newer minor versions.
```
Comment (abhayks1) [inline README.md]: Please update to ```stable version is released```

File: README.md [hunk 2]
```diff
@@ -183,3 +183,25 @@ The PR is the only place any team member is expected to consult for information
 [soliumrc]: ./.soliumrc.json
 [travisci]: https://travis-ci.org/
 [webpack]: https://webpack.js.org/
+
+
+## Managing internal and external dependencies
```
Comment (abhayks1) [inline README.md]: A new line after the heading "Managing internal and external dependencies" should be added.
Why it's ## and not ### ?

**Review by abhayks1 — DISMISSED** [review]
Nice work 💯 
Added couple of comments.

File: README.md (hunk 1 above)
Comment (abhayks1) [reply README.md]: What if the repo is new and no stable version is released?
e.g. 0.0.2 is the new minor version released after first minor version 0.0.1.
We should describe this case.

Comment (gulshanvasnani) [reply README.md]: I think you mean `patch` versions get released. So if the repo is new and patch version are being released then we should use tilde `~` to specify the package. 
Added it under internal dependencies with example.

Comment (gulshanvasnani) [reply README.md]: I think you mean `patch` versions get released. So if the repo is new and patch version are being released then we should use tilde `~` to specify the package. 
Added it under internal dependencies with example.

File: README.md [hunk 3]
```diff
@@ -183,3 +183,31 @@ The PR is the only place any team member is expected to consult for information
...
+
+### Managing internal and external dependencies
+* Internal dependencies:
+    1. Internal dependencies are dependencies which are under `@openst` project. 
+    2. Until the stable version is released, we should use version ranges to avoid any breaking change in newer minor versions.
+        Example:
+        ```
+               "dependencies": {
+                  "@openst/mosaic.js": ">=0.10.0 < 0.11.0"
+                }
```
Comment (schemar) [inline README.md]: Why are you indenting excessively again? Please fix.

File: README.md [hunk 4]
```diff
@@ -183,3 +183,31 @@ The PR is the only place any team member is expected to consult for information
...
+        Example:
+        ```
+               "dependencies": {
+                  "@openst/mosaic.js": ">=0.10.0 < 0.11.0"
+                }      
+        ``` 
+    3. When patch versions are released, then we should use tilde `~`.
+        ```
+               "dependencies": {
+                          "@openst/mosaic.js": "~0.1.0"
```
Comment (schemar) [inline README.md]: This doesn't make any sense. You can also use the tilde operator in the example above, but how is it different here than the above `">=0.10.0 < 0.11.0"`?

The tilde operator could be a good idea actually (to replace the above statement, not in addition), but see review comment and read documentation. Caret is actually sufficient.

**Review by schemar — CHANGES_REQUESTED** [review]
There seems to be a misunderstanding about what major/minor/patch versions are and what the difference between the various version range operators in npm is. See [advanced range syntax] in the official documentation.

So in our case, we should actually *always* use the caret version. And add a note to the readme why that also satisfies `0.x.x` releases in our case.

[advanced range syntax]: https://github.com/npm/node-semver#advanced-range-syntax

File: README.md (hunk 4 above)
Comment (gulshanvasnani) [reply README.md]: I went through the documentation mentioned.
1. We can use caret until stable version is released. Currently, we have `0.10.0` version released. So use of caret will allow auto-update between the range `>=0.10.0 < 0.11.0`.
2. Once we have stable version, use of caret will lead to updating to a version having breaking changes. Consider we have released `1.2.3` version. Use of caret will allow auto-update to latest version between `>=1.2.3 <2.0.0`. There might be some breaking changes in the minor versions which we have to avoid to get propagated.

Let me know your thoughts.

Comment (schemar) [reply README.md]: Very good 👍 
Once we have a stable 1.0.0 release, as per definition of [SemVer], there should not be any breaking changes in MINOR upgrades 😉 

So I think we can actually *always* use the caret for internal dependencies.

[semver]: https://semver.org/

File: README.md [hunk 5]
```diff
@@ -183,3 +183,24 @@ The PR is the only place any team member is expected to consult for information
 [soliumrc]: ./.soliumrc.json
 [travisci]: https://travis-ci.org/
 [webpack]: https://webpack.js.org/
+
+
+### Managing internal and external dependencies
```
Comment (schemar) [inline README.md]: Almost done 🎢 
Please move the entire section at the end of the `package.json` section.

**Review by schemar — CHANGES_REQUESTED** [review]

**Comment by schemar** [comment]
Moving back to "bounced". @gulshanvasnani please address the comments before moving the ticket to "review".

**Comment by schemar** [comment]
Moving back to "bounced" **again.** @gulshanvasnani **please** address the comments before moving the ticket to "review".

File: README.md [hunk 6]
```diff
@@ -140,6 +140,24 @@ This section applies to Solidity and JavaScript projects alike.
 * Define `files` to minimize the size of the published package.
 * For library projects, `main` should point to `./lib/index.js`.
 
+## Managing internal and external dependencies
```
Comment (schemar) [inline README.md]: Why did you change the header to be level 2??? Please change it back to be a level 3 header so that it is under the section "package.json".

```suggestion
### Managing internal and external dependencies
```

**Review by schemar — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (gulshanvasnani) [reply README.md]: I misunderstood the comment.
Reverted the change. Thank you.

**Review by schemar — APPROVED** [review]
:shipit:
